### ISSUE TITLE
Script micro optimisations

### ISSUE THREAD

**Issue by ajtowns** [jonasnick/GreatRSI#1]
https://github.com/jonasnick/GreatRSI/blob/664be9d7f0de4df5c56e91c5a1f8c13ccde12eec/GreatRSI/ScriptBuilder.lean#L50-L55

I think this should be `1ADD <G> TUCK SWAP CAT SWAP CHECKSIGVERIFY` rather than pushing G twice?

**Comment by jonasnick** [comment]
Oh, yes, thanks!

**Comment by ajtowns** [comment]
Oh, could also do a cumulative hash of the calculated pubkeys and only check the result, to save ~1900 bytes of script code.

**Comment by jonasnick** [comment]
Ah, nice idea! So instead of computing each `expected_pk[i]` and comparing it with `pk[i]`, you'd compute `hash(expected_pk[0], ..., expected_pk[n-1])` and compare with some `hash(pk[0], ..., pk[n-1]`, which you can obviously compute outside the script.

**Comment by ajtowns** [comment]
Yeah. Same suggestion is made at https://conduition.io/cryptography/quantum-hbs/#Modifications-1 fwiw. Or something like `pksum[0] = sha256(expected_pk[0]); pksum[n+1] = sha256(expected_pk[n+1] ++ pksum[n])` if you don't have streaming sha, and prefer computation to stack space usage.

**Comment by ajtowns** [comment]
Generating the randomizers by hashing the seed should also be reasonable, as far as I can see? Can't avoid hardcoding the seed and the pksum separately though, afaics.

**Comment by jonasnick** [comment]
> Same suggestion is made at https://conduition.io/cryptography/quantum-hbs/#Modifications-1 fwiw.

Thanks I didn't know about that blog post.

> Generating the randomizers by hashing the seed should also be reasonable, as far as I can see?

My immediate thought was that if it was reasonable, why wasn't it done in the [W-OTS+ paper](https://eprint.iacr.org/2017/965.pdf) or RFC 8391.

I had a superficial look at the proof of the W-OTS+ paper and it seems to me that generating the randomizers as hashes from public inputs would not work. The proof relies on manipulating the randomizers in a way that is indistinguishable for the adversary from honestly chosen randomizers (i.e., drawn uniformly at random). However, if `randomizer[i] = SHA256(seed,i)`, then the adversary could easily check whether it's seeing a regular or manipulated randomizer and just abort in the latter case.

However, I'm not really sure about the purpose of the randomizers in our case. It seems that they allow that the security of W-OTS+ is not relying on the collision resistance of the hash function. But we're relying on the collision resistance of SHA256 anyway when computing the transaction sighash. Note how the W-OTS+ paper assumes exactly n-byte messages. RFC 8391 uses a "randomized message hashing/digest" to prevent relying on collision resistance of the message hash. As far as I can see, the randomized hash is essentially just `hash(r, m)`, where `r` is unknown to the adversary.

IIUC, if we don't care about the collision resistance of SHA256, we can get rid of the randomizers. If we do care, then we need to consider in what scenarios sighashing is randomized. There's this [classic paper](https://cr.yp.to/hash/collisioncost-20090517.pdf) that shows that quantum computers are not better at finding collisions (and I don't know a rebuttal of that). But OTOH RFC 8391 did seem to consider that quantum computers may be better at collisions.

**Comment by ajtowns** [comment]
> I had a superficial look at the proof of the W-OTS+ paper and it seems to me that generating the randomizers as hashes from public inputs would not work. The proof relies on manipulating the randomizers in a way that is indistinguishable for the adversary from honestly chosen randomizers

I don't think that makes sense? The manipulated randomizer is used in a proof-by-contradiction to leverage a successful attacker of the protocol into an attack against the hash function; it seems fine for the manipulated randomizer to be out of protocol?

It looks to me like RFC 8391 section 4.1.4 describes generating the randomizers (BM_0 and BM_1) by hashing the SEED, fwiw, and section 4.1.7 seems to describe having the public key be just the two elements, what we called `pksum` above (calculated as a merkle root) and SEED? Couldn't guarantee I'm reading that correctly though.

**Comment by jonasnick** [comment]
> it seems fine for the manipulated randomizer to be out of protocol?

Can you elaborate? If the randomizer is out of the protocol and the adversary can compute it by themselves, it cannot be manipulated.

> It looks to me like RFC 8391 section 4.1.4 describes generating the randomizers (BM_0 and BM_1) by hashing the SEED

I would expect that the input to the hash algorithm includes some data that is secret to the verifier.

**Comment by jonasnick** [comment]
> I think this should be 1ADD <G> TUCK SWAP CAT SWAP CHECKSIGVERIFY rather than pushing G twice?

Cherry-picked your commit to use this shorter variant.

> Oh, could also do a cumulative hash of the calculated pubkeys and only check the result, to save ~1900 bytes of script code.

I hacked this into the code.

**Comment by ajtowns** [comment]
> > it seems fine for the manipulated randomizer to be out of protocol?
> 
> Can you elaborate? If the randomizer is out of the protocol and the adversary can compute it by themselves, it cannot be manipulated.

If I say "pi is 3.14", then you can manipulate that as "tau = 2*pi = 6.28". In the proof this is `r` and `r'`. I guess the problem with that is that the adversary `A` is assumed to be able to run an attack on any pubkey/seed/randomizer combination, whereas if the seed/randomizer have a known relationship to each other, you could have a successful adversary that's not that general.

> > It looks to me like RFC 8391 section 4.1.4 describes generating the randomizers (BM_0 and BM_1) by hashing the SEED
> 
> I would expect that the input to the hash algorithm includes some data that is secret to the verifier.

The verifier doesn't have any secrets, and has to run the hash algorithm to verify the signature, so I don't see how that makes sense? 4.1.10 specifies passing SEED in as a parameter to the verification algorithm fwiw.

I suppose one way you could make the randomizer be cheaply expressible yet still potentially completely arbitrary would be to include a function to generate the randomizer in the signature -- if you generate the randomizers from a seed, that's cheap; if you just list them out, that requires an arbitrarily powerful attacker. Pretty sure you could express that in simplicity or bll, not sure what the current thinking for GSR is re: OP_EVAL type things. Probably almost everyone would just choose the cheap option and it would just be security theater, though...

**Comment by jonasnick** [comment]
> The verifier doesn't have any secrets

Sure. I said "data that is secret **to** the verifier", i.e., some data that only the signer knows.

I've had a closer look at the scheme specified in the RFC and such data in fact does not exist. Randomizers are computed by applying a PRF with a public seed to a public input. This is already evident from the chaining pseudocode:

```
ADRS.setHashAddress(i + s - 1);
ADRS.setKeyAndMask(0);
KEY = PRF(SEED, ADRS);
ADRS.setKeyAndMask(1);
BM = PRF(SEED, ADRS);

tmp = F(KEY, tmp XOR BM);
```

`SEED` is part of the public key (as in bare W-OTS+) and `ADRS` really is just a public, low entropy value that ensures the PRF is evaluated on a fresh input.

In fact, the first version of the RFC (`00`) used explicit bitmasks that were part of the public key, but this was changed in version `01`.

Why do they use a scheme that is obviously not covered by the W-OTS+ security proof? I found the answer in the paper [Rapidly Verifiable XMSS Signatures](https://tches.iacr.org/index.php/TCHES/article/download/8730/8330/5451) that was published 3 years after the RFC was finalized (and is not cited in the RFC):

> XMSS in RFC 8391 uses a scheme commonly referred to as WOTS+ (which we follow, although it actually is WOTS-T [[HRS16b](https://www.iacr.org/archive/pkc2016/96140179/96140179.pdf)])

Following the citation, we find the variant of WOTS that computes randomizers from public inputs and a security proof of that variant. This proof requires the (quantum) random oracle model (a stronger model than used in the original WOTS+ scheme), which is programmed to output a "manipulated" randomizer value.

## tl;dr

We should generate each randomizer `i` as `<SEED> <i> OP_CAT OP_SHA256`.

## Appendix

> If I say "pi is 3.14", then you can manipulate that as "tau = 2*pi = 6.28". In the proof this is r and r'.

This is not a helpful comparison. In WOTS+ the randomizer is chosen uniformly at random and we have a security proof that requires setting that randomizer to a specific value in order to extract a second preimage from the adversaries forgery. As explicitly mentioned in the proof, the adversary can not distinguish that value from a value that is drawn honestly from the uniform distribution.

If we change WOTS+ to a scheme where the randomizer is output of a PRF instead of the public key. Then the security proof will simply not apply: The adversary computes the randomizer using the PRF and with overwhelming probability it will not be the specific value that would allow extracting the second preimage.

Hence, the theorem and security proof needs to be adapted by replacing the PRF with a random oracle. The random oracle can then be programmed by the reduction to output the specific value when the adversary computes the signers randomizer. Apparently the WOTS-T paper demonstrates that this is possible.

**Comment by ajtowns** [comment]
> Sure. I said "data that is secret **to** the verifier", i.e., some data that only the signer knows.

Ah, I read that exactly backwards: If you keep a secret from someone, that means they don't know it; surely therefore a secret to someone means the opposite and they do know it. *shrug*

I think we're on the same page now; and are able to reduce the data encoded into the script by ~2kB by reducing pubkeys and another ~0.5kB by reducing randomization info?

**Comment by jonasnick** [comment]
> ~0.5kB by reducing randomization info?

Yeah.

> ~2kB by reducing pubkeys

How do we reduce pubkeys further?

**Comment by ajtowns** [comment]
> > ~2kB by reducing pubkeys
> 
> How do we reduce pubkeys further?

Not further, that's just this with perhaps some rounding:

> > Oh, could also do a cumulative hash of the calculated pubkeys and only check the result, to save ~1900 bytes of script code.
> 
> I hacked this into the code.

**Comment by jonasnick** [comment]
I'm tempted to remove the randomizers entirely. If collision resistance of SHA256 is broken, we have bigger problems, such as chain splits.

**Comment by jonasnick** [comment]
While reading https://eprint.iacr.org/2025/055.pdf I came across another possible optimization: "Target Sum Winternitz", where the checksum is fixed by the verifier instead of being an explicit part of the signature, at the cost of additional signing work.

In this scheme, the signer has to hash their message with some randomness, encode into chunks and check if the sum of the chunks matches the checksum. If not, they rehash the message with new randomness until they have found the randomness that results in the correct checksum.
